### ISSUE TITLE
Don't allow regular tabs to position themselves within pinned tab groups

### ISSUE THREAD

**Issue by cryzed** [qutebrowser/qutebrowser#2898]
Maybe I'm too used to how Firefox did things, but its tab-handling made sense to me. I pin tabs that I frequent and don't close: they are a group of tabs I can quickly navigate to.

In qutebrowser it's currently possible to "break up" a pinned group of tabs, by either shifting the tab manually (`gr`, `gl`...) or by focusing on a pinned tab and opening a link in a new tab non-explicitly (using `F` or `;r`).

I would expect the tab that I'm creating to either appear on the "right side" completely outside of the pinned tab group, and tabs that I attempt to shift to the left side of it, to _either_ stop in front of the pinned tabs (fixing the pinned tab group the very left side of all tabs) _or_ appear to the left side of the pinned tabs.

I thought some discussion on this would probably be best, because in the end it's a subjective design-decision.

**Comment by The-Compiler** [comment]
cc @jgkamat :wink:

**Comment by parchd-1** [comment]
> I would expect the tab that I'm creating to either appear on the "right side" completely outside of the pinned tab group

This is the behaviour I expected when I started to use pinned tabs in qutebrowser, and I've never used them in Firefox or any other browser before.

**Comment by jgkamat** [comment]
When I initially designed pinned tabs, I tried to make them as simple as possible, so that people can build a 'pinned tab' workflow for themselves. I don't think that's possible until the plugin api is available though, and this behavior does seem a lot more intuitive for the most common use case of pinned tabs.

So far I'm seeing a few different 'knobs' we can tweak:

1. newly created tabs should follow `new-tab-position` (current) or they should open to the right of the tab group, if opening the tab would break up the tab group
2. Shifting an unpinned tab 'into' a tab group should break up the tab group, or it should move 'over' the tab group. 
3. Shifting a pinned tab 'out of' a tab group should cause it to leave the tab group, or cause it to wrap around to the end of the tab group (keeping it in the group).

I currently have a workflow where I maintain two groups of pinned tabs, one on the left (for long running tabs that I access frequently, in a queue) and a group on the right (for low priority tabs that I would like to return to in a LIFO order (like a stack).

I definelty would like `1`, maybe under an option to turn it on or off, as it would make opening links in pinned tabs a little bit more easy. On the right hand side, maybe tabs can open to the 'left' of the pinned tab group, to keep the group on the end.

I don't know if `2` or `3` is a good idea, because it would make it much harder to intentionally split up tab groups. For example, I regulary move pinned tabs from the left group to the right group and vice versa if they lower/raise in priority. I could imagine a vertical tab user having many tab groups, and it would be painful to move pinned tabs between groups if something like this was available. Thiis idea needs more thought though.

If anyone has any additional feedback, please let me know in this thread, I would like to try to keep pinned tabs as flexible as possible (at least as much as they are now), while improving the usability of them, if possible :smile: 

Also, this is probably going to be blocked on #2847, which is blocked on the new config :).

**Comment by cryzed** [comment]
What you describe in `1`-`3` seems to cover all possible scenarios when either regarding a group of pinned tabs as a cohesive unit or simply as regular tabs. Combining all behaviors you described into a single option, e.g. `cohesive-pinned-tabs` or something among those lines, would be the most simple and clean implementation - I think creating an option for each of these would be overkill and needlessly increase the complexity of the tab-ordering code and configuration for the end-user.

There are a few corner cases that are hard to figure out for example:

 * What happens if you remove all tabs between 2 groups of pinned tabs: are those merged into a single pinned tab group?
 * How do you define a new group of tabs? Simply pinning a single tab and those around it?
 * Do we indicate somehow that pinned tabs are part of a group?
 * Will it be possible to have pinned tabs next to each other that don't form a group?

To work around the "issue" currently, I simply decided to set `new-tab-position` to `last`, which works for me currently, although it's clearly different from the `next` option; however it works so well that I'm questioning my initial suggestion and if it's really worth implementing this with all the murky logic.

**Comment by cryzed** [comment]
I'll close this for now.

**Comment by The-Compiler** [comment]
I think it's fine to keep this open, it certainly needs some more thought.

**Comment by peacememories** [comment]
I, too, would love to see an option for this behaviour. I keep a list of pinned tabs on the left for things I *always* need during work, for example our chat, email, issue tracker. I often open tabs from these applications and then have to move them after the pinned group so I don't break it up. setting `new-tab-position` would be an option to fix this, but I actually want the `next` behaviour to keep logically connected tabs together.
